pkg/sql/tests/hash_sharded_test.go

```diff
@@ -0,0 +1,227 @@
+	shardColID := getShardColumnID(t, tableDesc, shardedIndexName)
+	foundCheckConstraint := false
+	for _, check := range tableDesc.AllActiveAndInactiveChecks() {
+		usesShard, err := check.UsesColumn(tableDesc, shardColID)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if usesShard && check.Hidden {
+			foundCheckConstraint = true
+			break
+		}
+	}
+	if !foundCheckConstraint {
+		t.Fatalf(`Could not find hidden check constraint for shard column`)
+	}
+	if idx.ColumnIDs[0] != shardColID {
+		t.Fatalf(`Expected shard column to be the first column in the set of index columns`)
+	}
+}
+
+func TestBasicHashShardedIndexes(t *testing.T) {
+	ctx := context.Background()
+	s, db, kvDB := serverutils.StartServer(t, base.TestServerArgs{})
+	defer s.Stopper().Stop(ctx)
+	if _, err := db.Exec(`CREATE DATABASE d`); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := db.Exec(`USE d`); err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run("primary", func(t *testing.T) {
+		if _, err := db.Exec(`
+			CREATE TABLE kv_primary (
+				k INT PRIMARY KEY USING HASH WITH BUCKET_COUNT=5,
+				v BYTES
+			)
+		`); err != nil {
+			t.Fatal(err)
+		}
+
+		if _, err := db.Exec(`CREATE INDEX foo ON kv_primary (v)`); err != nil {
+			t.Fatal(err)
+		}
+		tableDesc := sqlbase.GetTableDescriptor(kvDB, `d`, `kv_primary`)
+		verifyTableDescriptorState(t, tableDesc, "primary" /* shardedIndexName */)
+		shardColID := getShardColumnID(t, tableDesc, "primary" /* shardedIndexName */)
+
+		// Ensure that secondary indexes on table `kv` have the shard column in their
+		// `ExtraColumnIDs` field so they can reconstruct the sharded primary key.
+		fooDesc, _, err := tableDesc.FindIndexByName("foo")
+		if err != nil {
+			t.Fatal(err)
+		}
+		foundShardColumn := false
+		for _, colID := range fooDesc.ExtraColumnIDs {
+			if colID == shardColID {
+				foundShardColumn = true
+				break
+			}
+		}
+		if !foundShardColumn {
+			t.Fatalf(`Secondary index cannot reconstruct sharded primary key`)
+		}
+	})
+
+	t.Run("secondary_in_create_table", func(t *testing.T) {
+		if _, err := db.Exec(`
+			CREATE TABLE kv_secondary (
+				k INT,
+				v BYTES,
+				INDEX sharded_secondary (k) USING HASH WITH BUCKET_COUNT = 12
+			)
+		`); err != nil {
+			t.Fatal(err)
+		}
+
+		tableDesc := sqlbase.GetTableDescriptor(kvDB, `d`, `kv_secondary`)
+		verifyTableDescriptorState(t, tableDesc, "sharded_secondary" /* shardedIndexName */)
+	})
+
+	t.Run("secondary_in_separate_ddl", func(t *testing.T) {
+		if _, err := db.Exec(`
+			CREATE TABLE kv_secondary2 (
+				k INT,
+				v BYTES
+			)
+		`); err != nil {
+			t.Fatal(err)
+		}
+
+		if _, err := db.Exec(`CREATE INDEX sharded_secondary2 ON kv_secondary2 (k) USING HASH WITH BUCKET_COUNT = 12`); err != nil {
+			t.Fatal(err)
+		}
+		tableDesc := sqlbase.GetTableDescriptor(kvDB, `d`, `kv_secondary2`)
+		verifyTableDescriptorState(t, tableDesc, "sharded_secondary2" /* shardedIndexName */)
+	})
+}
+
+func TestHashShardedIndexVersionGating(t *testing.T) {
```

Comment:
I'm not sure we need this test case -- I have a PR landing soon (https://github.com/cockroachdb/cockroach/pull/43332) that adds a mixed 19.2-20.1 cluster setting to the logic tests, so we could just add a test that ensures you can't create a hash sharded index in that state.